File: ambeth/Ambeth.Merge.Test/ambeth/merge/DefaultProxyHelper.cs
using De.Osthus.Ambeth.Log;
#if !SILVERLIGHT
using Castle.DynamicProxy;
#else
using Castle.Core.Interceptor;
#endif
using System;
using System.Collections;
using De.Osthus.Ambeth.Merge.Model;
using De.Osthus.Ambeth.Typeinfo;
using De.Osthus.Ambeth.Cache;

namespace De.Osthus.Ambeth.Merge
{
    public class DefaultProxyHelper : IProxyHelper
    {
        [LogInstance]
		public ILogger Log { private get; set; }

        public virtual Type GetRealType(Type type)
        {
            if (typeof(IProxyTargetAccessor).IsAssignableFrom(type))
            {
                Type baseType = type.BaseType;
                if (typeof(Object).Equals(baseType))
                {
                    return type.GetInterfaces()[0];
                }
                return baseType;
            }
            return type;
        }

        public bool ObjectEquals(Object leftObject, Object rightObject)
        {
            if (leftObject == null)
            {
                return rightObject == null;
            }
            if (rightObject == null)
            {
                return false;
            }
            if (leftObject == rightObject)
            {
                return true;
            }
            return Object.Equals(leftObject, rightObject);
        }

        public bool IsInitialized(Object parentObj, String memberName)
        {
            throw new NotSupportedException();
        }

        public bool IsInitialized(Object parentObj, IRelationInfoItem member)
        {
            throw new NotSupportedException();
        }

        public void SetUninitialized(Object parentObj, IRelationInfoItem member, IObjRef[] objRefs)
        {
            throw new NotSupportedException();
        }

        public IObjRef[] GetObjRefs(Object parentObj, String memberName)
        {
            throw new NotSupportedException();
        }

        public IObjRef[] GetObjRefs(Object parentObj, IRelationInfoItem member)
        {
            throw new NotSupportedException();
        }

        public void SetObjRefs(Object parentObj, String memberName, IObjRef[] objRefs)
        {
            throw new NotSupportedException();
        }

        public void SetObjRefs(Object parentObj, IRelationInfoItem member, IObjRef[] objRefs)
        {
            throw new NotSupportedException();
        }

        public object GetValueDirect(Object parentObj, IRelationInfoItem member)
        {
            throw new NotSupportedException();
        }

        public void SetInitPending(Object parentObj, IRelationInfoItem member)
        {
            throw new NotSupportedException();
        }

        public void SetValueDirect(Object parentObj, IRelationInfoItem member, Object value)
        {
            throw new NotSupportedException();
        }

        public void SetInitialized(Object parentObj, IRelationInfoItem member, Object value)
        {
            throw new NotSupportedException();
        }

        public bool GetInitPending(Object parentObj, IRelationInfoItem member)
        {
            throw new NotSupportedException();
        }

        public ValueHolderState GetState(Object parentObj, IRelationInfoItem member)
        {
            throw new NotSupportedException();
        }
    }
}

File: jambeth/jambeth-merge/src/test/java/de/osthus/ambeth/merge/DefaultProxyHelper.java
package de.osthus.ambeth.merge;

import de.osthus.ambeth.bytecode.IBytecodeEnhancer;
import de.osthus.ambeth.ioc.IInitializingBean;
import de.osthus.ambeth.ioc.annotation.Autowired;
import de.osthus.ambeth.log.ILogger;
import de.osthus.ambeth.log.LogInstance;
import de.osthus.ambeth.proxy.ICgLibUtil;

public class DefaultProxyHelper implements IProxyHelper, IInitializingBean
{
	@SuppressWarnings("unused")
	@LogInstance
	private ILogger log;

	@Autowired
	protected ICgLibUtil cgLibUtil;

	@Autowired(optional = true)
	protected IBytecodeEnhancer entityEnhancer;

	@Override
	public void afterPropertiesSet() throws Throwable
	{
		// Intended blank
	}

	@Override
	public Class<?> getRealType(Class<?> type)
	{
		IBytecodeEnhancer entityEnhancer = this.entityEnhancer;
		if (entityEnhancer != null)
		{
			Class<?> baseType = entityEnhancer.getBaseType(type);
			if (baseType != null)
			{
				return baseType;
			}
		}
		return cgLibUtil.getOriginalClass(type);
	}

	@Override
	public boolean objectEquals(Object leftObject, Object rightObject)
	{
		if (leftObject == null)
		{
			return rightObject == null;
		}
		if (rightObject == null)
		{
			return false;
		}
		if (leftObject == rightObject)
		{
			return true;
		}
		return leftObject.equals(rightObject);
	}
}
